src/formats/GRO.cpp

```diff
@@ -0,0 +1,324 @@
+// Chemfiles, a modern library for chemistry file reading and writing
+// Copyright (C) Guillaume Fraux and contributors -- BSD license
+
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include "chemfiles/formats/GRO.hpp"
+
+#include "chemfiles/ErrorFmt.hpp"
+#include "chemfiles/File.hpp"
+#include "chemfiles/Frame.hpp"
+#include "chemfiles/utils.hpp"
+#include "chemfiles/warnings.hpp"
+
+using namespace chemfiles;
+
+template<> FormatInfo chemfiles::format_information<GROFormat>() {
+    return FormatInfo("GRO").with_extension(".gro").description(
+        "GROMACS GRO text format"
+    );
+}
+
+/// Check the number of digits before the decimal separator to be sure than
+/// we can represen them. In case of error, use the given `context` in the error
+/// message
+static void check_values_size(const Vector3D& values, unsigned width, const std::string& context);
+/// Fast-forward the file for one step, returning `false` if the file does
+/// not contain one more step.
+static bool forward(TextFile& file);
+
+GROFormat::GROFormat(const std::string& path, File::Mode mode)
+    : file_(TextFile::create(path, mode))
+{
+    while (!file_->eof()) {
+        auto position = file_->tellg();
+        if (!file_ || position == std::streampos(-1)) {
+            throw format_error("IO error while reading '{}' as GRO", path);
+        }
+        if (forward(*file_)) {
+            steps_positions_.push_back(position);
+        }
+    }
+    file_->rewind();
+}
+
+size_t GROFormat::nsteps() {
+    return steps_positions_.size();
+}
+
+void GROFormat::read_step(const size_t step, Frame& frame) {
+    assert(step < steps_positions_.size());
+    file_->seekg(steps_positions_[step]);
+    read(frame);
+}
+
+void GROFormat::read(Frame& frame) {
+    long count = 0;
+    try {
+        frame.set("name", (file_->readline())); // GRO comment line;
+        count = std::stol(file_->readline());
+    } catch (const std::exception& e) {
+        throw format_error("can not read next step as GRO: {}", e.what());
+    }
+
+    if (count < 0) {
+        throw format_error("the number of atoms can not be negative in GRO format");
+    }
+
+    size_t natoms = static_cast<size_t>(count);
+
+    residues_.clear();
+    frame.add_velocities();
+    frame.reserve(natoms);
+    frame.resize(0);
+
+    for (const auto& line: file_->readlines(natoms)) {
+        auto resid = std::stoul(line.substr(0,5));
```

Comment:
could you add a check that the line is long enough?

Reply:
Fair enough!